### ISSUE TITLE
Error at login: AnonymousUser

### ISSUE THREAD

**Issue by sposs** [WeblateOrg/weblate#1610]
### Steps to reproduce
Upgrade using the git repository
Logout from admin page

### Actual behaviour
Server error: AnonymousUser is not iterable

### Expected behaviour
Login page should be displayed

The problem comes from the fact that the all_acl function in trans/models/project.py tries to return project ACLs for logged out users, using therefore the Anonymous user which has no ACLs.

I did run the setupgroups command after the update.

For reference, here is the relevant bit in the log (line 66):

File "/usr/share/weblate/weblate/trans/context_processors.py" in weblate_context
  43.     projects = Project.objects.all_acl(request.user)

File "/usr/share/weblate/weblate/trans/models/project.py" in all_acl
  81.         return self.filter(id__in=self.get_acl_ids(user))

File "/usr/share/weblate/weblate/trans/models/project.py" in get_acl_ids
  66.                 groupacl__groups__user=user,

File "/usr/local/lib/python2.7/dist-packages/django/db/models/manager.py" in manager_method
  85.                 return getattr(self.get_queryset(), name)(*args, **kwargs)

File "/usr/local/lib/python2.7/dist-packages/django/db/models/query.py" in filter
  796.         return self._filter_or_exclude(False, *args, **kwargs)

File "/usr/local/lib/python2.7/dist-packages/django/db/models/query.py" in _filter_or_exclude
  814.             clone.query.add_q(Q(*args, **kwargs))

File "/usr/local/lib/python2.7/dist-packages/django/db/models/sql/query.py" in add_q
  1227.         clause, _ = self._add_q(q_object, self.used_aliases)

File "/usr/local/lib/python2.7/dist-packages/django/db/models/sql/query.py" in _add_q
  1253.                     allow_joins=allow_joins, split_subq=split_subq,

File "/usr/local/lib/python2.7/dist-packages/django/db/models/sql/query.py" in build_filter
  1158.             self.check_related_objects(field, value, opts)

File "/usr/local/lib/python2.7/dist-packages/django/db/models/sql/query.py" in check_related_objects
  1057.                 for v in value:

File "/usr/local/lib/python2.7/dist-packages/django/utils/functional.py" in inner
  235.         return func(self._wrapped, *args)

I fixed it in my version by checking if the user is authenticated before attempting to get the ACLs.

**Comment by sposs** [comment]
to be complete
``` python
   def all_acl(self, user):
        """Return list of projects user is allowed to access
        and flag whether there is any filtering active.
        """
        if not user.is_authenticated:
            return []
        if user.is_superuser:
            return self.all()
        return self.filter(id__in=self.get_acl_ids(user))
```

**Comment by nijel** [comment]
This should have been fixed as of 149a5d7f7c56194de64d5c8edc7e69ed16251437, see #1444.

Do you have properly installed Weblate Admin module?

**Comment by sposs** [comment]
I have followed the upgrade instructions. I was at 2.10, am now at 2.17. Are there things to do not mentionned in the upgrade instructions?

**Comment by nijel** [comment]
Most likely some of the changes in INSTALLED APPS are missing (note the append/insert part at the end):

https://github.com/WeblateOrg/weblate/blob/master/weblate/settings_example.py#L327-L358

**Comment by sposs** [comment]
Oh, indeed, that's possible. I did check the settings_example.py file, but I may have missed that. Will check tomorrow.

**Comment by nijel** [comment]
Well it was changed in https://github.com/WeblateOrg/weblate/pull/1597 in order to make it harder to mess it up ;-)

**Comment by sposs** [comment]
That was indeed something I had overlooked. I already have my tweaked settings file, so I had to port the changes manually given what was written in the migration doc, and I missed that part. 

In any case, if that solves the initial problem, I have a problem with the logic: AFAIK apps' orders should not matter with Django... Also, in that particular problem, why do you look up ACLs for anonymous users? Does it make sense?

Thanks anyway for the hints.

**Comment by nijel** [comment]
Order does matter as that's order in which templates are looked up and we need to override some of the Django admin templates, that's why the Weblate admin app needs to be in front of Django admin app (it really doesn't have to be first, but that's just simplification). Having weblate app last is way to detect last post_migrate signal when setting up Weblate groups, this could be probably done without relying on app ordering though...

The anonymous user is real user in Weblate compared to Django where it's just a virtual one not stored in database. This is needed to be able to change permissions for it - for example what anonymous user is able to see.

**Comment by sposs** [comment]
Then I would add in the migration step where the order becomes important that the settings file must be fixed before running the migration, therefore the Generic upgrade instructions are to be done after that. I certainly did not, as I followed the order shown. This lead to complications. 

I have been struggling for 2h with the permissions (group ACLs) because currently the doc is not clear at all to me. None of my projects have Enable ACL on, but the group ACLs are still used, which is not logical to me, and the setupgroups command should not create ACLs for projects that do not use ACLs because it puts the system in a very strange state. I had to remove all those default group ACLs and recreate them by hand. You can certainly feel the frustration in my wording... Nice project, but it becomes too large compared to the doc. For example there are too many permissions and those are not always described in the doc.

**Comment by nijel** [comment]
I've just fixed one of the needs for ordering in 871bfc9fd, I'll look into another (that should not be hard as well).

The group ACLs are used to define project admins. The generic Django permissions do not allow per object control. I know the permissions are way more complex for most use cases, however this is pretty much outcome of being able to handle more complex setups some people use...

In most cases you really do not need to touch them at all directly and do the management from Weblate (see https://docs.weblate.org/en/latest/admin/access.html#managing-per-project-access-control).

**Comment by nijel** [comment]
Thanks for pushing me into this, the app ordering is no longer important.
